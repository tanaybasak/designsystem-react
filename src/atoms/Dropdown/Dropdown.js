import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import prefix from '../../settings';

const Dropdown = ({ type, items, id, label }) => {
    const [isOpen, setIsOpen] = useState(false);
    const [selected, setSelected] = useState(null);
    const [typeState, setTypeState] = useState(type);

    useEffect(() => {
        positionDropDown()
    });

    const onSelect = (event) => {
        setIsOpen(false);
        setSelected(event.target.innerText);
    };

    const positionDropDown = () => {
        let getBound;
        if (document.getElementById(id).getElementsByTagName('ul')[0]) {
            getBound = document.getElementById(id).getElementsByTagName('ul')[0].getBoundingClientRect();
            // To open bottom dropdown at top
            if (window.innerHeight < getBound.bottom && type === 'bottom') {
                setTypeState('top')
            }
            // To open top dropdown at bottom
            if (getBound.top + window.pageYOffset < 0 && type === 'top') {
                setTypeState('down')
            }
        }
    }
    return (
        <section className={`${prefix}-dropdown ${typeState === "bottom" ? `${prefix}-dropdown-bottom` : `${prefix}-dropdown-top`}
             ${isOpen ? `${prefix}-dropdown-open` : ""}`} data-component="dropdown" id={id}
        >
            <button className={`${prefix}-btn ${prefix}-dropdown-toggle`}
                data-toggle="dropdown"
                onClick={() => { setIsOpen(true) }}
            >
                {selected ? selected : label}
            </button>
            {
                isOpen
                    ?
                    <ul className={`${prefix}-dropdown-container`} aria-labelledby="dropdownMenuButton">
                        {
                            items.map(item => {
                                return (<li className={`${prefix}-dropdown-item`} key={item.id} onClick={onSelect}>{item.text}</li>)
                            })
                        }
                    </ul>
                    : null}
        </section>
    );
};

Dropdown.propTypes = {
    type: PropTypes.string,
    items: PropTypes.array.isRequired,
    id: PropTypes.string.isRequired,
    label: PropTypes.string
};

Dropdown.defaultProps = {
    type: "down",
    label: "Select Option"
};

export default Dropdown;
