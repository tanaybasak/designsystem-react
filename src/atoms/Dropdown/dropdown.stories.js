import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { boolean, object, text } from '@storybook/addon-knobs';
//@update-path-build-start
import Dropdown from './Dropdown';
//@update-path-build-end

const items = [
  {
    id: 'option-1',
    text: 'Option 1'
  },
  {
    id: 'option-2',
    text: 'Option 2'
  },
  {
    id: 'option-3',
    text: 'Option 3'
  },
  {
    id: 'option-4',
    text: 'Option 4'
  }
];

storiesOf('Dropdown', module)
  .add(
    'default',
    () => (
      <Dropdown
        items={object('Items', items)}
        label={text('Label', 'Dropdown Label')}
        onChange={action('Dropdown-Onchange')}
        disabled={boolean('disabled', false)}
      />
    ),
    {
      info: {
        text: `Description About Dropdown Component \n
        import { Dropdown } from '@patron/patron-react/dropdown';`
      }
    }
  )
  .add(
    'top',
    () => (
      <Dropdown
        type={'top'}
        items={object('Items', items)}
        label={text('Label', 'Dropdown Label')}
        selectedItem={text('Item', 'option-2')}
        onChange={action('Dropdown-Onchange')}
        disabled={boolean('disabled', false)}
      />
    ),
    {
      info: {
        text: `Description About Dropdown Component \n
        import { Dropdown } from '@patron/patron-react/dropdown';`
      }
    }
  )
  .add(
    'multselect',
    () => (
      <Dropdown
        dropdownType="multi"
        items={object('Items', items)}
        label={text('Label', 'MultiSelect Label')}
        onChange={action('Dropdown-Onchange')}
        disabled={boolean('disabled', false)}
      />
    ),
    {
      info: {
        text: `Description About Dropdown Component \n
        import { Dropdown } from '@patron/patron-react/dropdown';`
      }
    }
  );
