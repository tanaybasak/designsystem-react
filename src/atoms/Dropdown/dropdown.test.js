import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import prefix from '../../settings';
import Dropdown from './Dropdown';

const items = [
  {
    id: 'option-1',
    text: 'Option 1'
  },
  {
    id: 'option-2',
    text: 'Option 2'
  },
  {
    id: 'option-3',
    text: 'Option 3'
  },
  {
    id: 'option-4',
    text: 'Option 4'
  }
];

describe('<Dropdown> component', () => {
  it('snapshots/renders Dropdown correctly', () => {
    const tree = renderer.create(<Dropdown items={items} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('triggers click event in Dropdown', () => {
    const mockonclick = jest.fn();
    const dropdown = mount(<Dropdown items={items} onClick={mockonclick} />);
    dropdown.find(`.${prefix}-dropdown-toggle`).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeTruthy();
  });

  it('Arrow Up event on Dropdown button', () => {
    const mockonclick = jest.fn();
    const dropdown = mount(<Dropdown items={items} onClick={mockonclick} />);
    dropdown.find(`.${prefix}-dropdown-toggle`).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeTruthy();

    let button = dropdown.find(`.${prefix}-dropdown-toggle`);
    button.simulate('keydown', { keyCode: 38 });

    expect(dropdown.find('li').at(3).is(':focus')).toBe(true);
    dropdown.find('li').at(3).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeFalsy();
    expect(button.text()).toEqual('Option 4');
  });

  it('Arrow Down event on Dropdown button', () => {
    const mockonclick = jest.fn();
    const dropdown = mount(<Dropdown items={items} onClick={mockonclick} />);
    dropdown.find(`.${prefix}-dropdown-toggle`).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeTruthy();

    let button = dropdown.find(`.${prefix}-dropdown-toggle`);
    button.simulate('keydown', { keyCode: 40 });

    expect(dropdown.find('li').at(0).is(':focus')).toBe(true);
    dropdown.find('li').at(0).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeFalsy();
    expect(button.text()).toEqual('Option 1');
  });

  it('Navigate Item using Arrow Down', () => {
    const mockonclick = jest.fn();
    const dropdown = mount(<Dropdown items={items} onClick={mockonclick} />);
    dropdown.find(`.${prefix}-dropdown-toggle`).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeTruthy();

    let button = dropdown.find(`.${prefix}-dropdown-toggle`);

    button.simulate('keydown', { keyCode: 40 });

    let anchor = dropdown.find('li').at(0);
    expect(anchor.is(':focus')).toBe(true);

    anchor.simulate('keydown', { keyCode: 40 });
    anchor = dropdown.find('li').at(1);

    expect(anchor.is(':focus')).toBe(true);

    anchor.simulate('keydown', { keyCode: 40 });
    anchor = dropdown.find('li').at(2);

    expect(anchor.is(':focus')).toBe(true);

    anchor.simulate('keydown', { keyCode: 40 });
    anchor = dropdown.find('li').at(3);

    expect(anchor.is(':focus')).toBe(true);
  });

  it('Navigate Item using Arrow Up', () => {
    const mockonclick = jest.fn();
    const dropdown = mount(<Dropdown items={items} onClick={mockonclick} />);
    dropdown.find(`.${prefix}-dropdown-toggle`).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeTruthy();

    let button = dropdown.find(`.${prefix}-dropdown-toggle`);

    button.simulate('keydown', { keyCode: 38 });

    let anchor = dropdown.find('li').at(3);
    expect(anchor.is(':focus')).toBe(true);

    anchor.simulate('keydown', { keyCode: 38 });
    anchor = dropdown.find('li').at(2);

    expect(anchor.is(':focus')).toBe(true);

    anchor.simulate('keydown', { keyCode: 38 });
    anchor = dropdown.find('li').at(1);

    expect(anchor.is(':focus')).toBe(true);

    anchor.simulate('keydown', { keyCode: 38 });
    anchor = dropdown.find('li').at(0);

    expect(anchor.is(':focus')).toBe(true);
  });

  it('Navigate Item using Arrow Up multiSelectDropdown', () => {
    const mockonclick = jest.fn();
    const dropdown = mount(
      <Dropdown items={items} dropdownType="multi" onClick={mockonclick} />
    );
    dropdown.find(`.${prefix}-dropdown-toggle`).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeTruthy();

    let button = dropdown.find(`.${prefix}-dropdown-toggle`);

    button.simulate('keydown', { keyCode: 38 });

    let item = dropdown.find('li').at(3);
    expect(item.is(':focus')).toBe(true);

    item.simulate('keydown', { keyCode: 38 });
    item = dropdown.find('li').at(2);

    expect(item.is(':focus')).toBe(true);

    item.simulate('keydown', { keyCode: 38 });
    item = dropdown.find('li').at(1);

    expect(item.is(':focus')).toBe(true);

    item.simulate('keydown', { keyCode: 38 });
    item = dropdown.find('li').at(0);

    expect(item.is(':focus')).toBe(true);
  });

  it('Navigate Item using Arrow Down multiSelectDropdown', () => {
    const mockonclick = jest.fn();
    const dropdown = mount(
      <Dropdown items={items} dropdownType="multi" onClick={mockonclick} />
    );
    dropdown.find(`.${prefix}-dropdown-toggle`).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeTruthy();

    let button = dropdown.find(`.${prefix}-dropdown-toggle`);

    button.simulate('keydown', { keyCode: 40 });

    let item = dropdown.find('li').at(0);
    expect(item.is(':focus')).toBe(true);

    item.simulate('keydown', { keyCode: 40 });
    item = dropdown.find('li').at(1);

    expect(item.is(':focus')).toBe(true);

    item.simulate('keydown', { keyCode: 40 });
    item = dropdown.find('li').at(2);

    expect(item.is(':focus')).toBe(true);

    item.simulate('keydown', { keyCode: 40 });
    item = dropdown.find('li').at(3);

    expect(item.is(':focus')).toBe(true);
  });

  it('select MultiSelect Dropdown item', () => {
    const mockonclick = jest.fn();
    const dropdown = mount(
      <Dropdown items={items} dropdownType="multi" onClick={mockonclick} />
    );
    dropdown.find(`.${prefix}-dropdown-toggle`).simulate('click');
    expect(dropdown.find(`.${prefix}-dropdown-menu`).exists()).toBeTruthy();

    let button = dropdown.find(`.${prefix}-dropdown-toggle`);
    button.simulate('keydown', { keyCode: 40 });

    expect(dropdown.find('li').at(0).is(':focus')).toBe(true);
    dropdown.find('li').at(0).simulate('click');
    const input = dropdown.find('input').at(0);
    expect(input.props().checked).toBe(true);
  });
});
