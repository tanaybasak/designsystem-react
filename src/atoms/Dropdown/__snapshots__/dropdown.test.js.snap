// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown> component snapshots/renders Dropdown correctly 1`] = `
<div
  className="hcl-overlay-wrapper hcl-dropdown"
>
  <button
    aria-haspopup="true"
    aria-label="Select Option"
    className="hcl-btn hcl-dropdown-toggle"
    data-toggle="dropdown"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    Select Option
  </button>
</div>
`;
